benchmarking/src/performance_evaluation.py

```diff
@@ -0,0 +1,860 @@
+            individual_responses (list): A list of individual responses from the performance evaluation.
+
+        Returns:
+            None
+
+        Raises:
+            ValueError: If the results directory does not exist or is not a directory.
+        """
+        filename = re.sub(r"[^\w\d-]+", "-", filename)
+        filename = re.sub(r"-{2,}", "-", filename)
+        summary_filename = f"{filename}_summary"
+        individual_responses_filename = f"{filename}_individual_responses"
+
+        # Update to metadata.
+        summary.update(self.user_metadata)
+
+        results = LLMPerfResults(name=summary_filename, metadata=summary)
+        results_dir = Path(self.results_dir)
+        if not results_dir.exists():
+            results_dir.mkdir(parents=True)
+        elif not results_dir.is_dir():
+            raise ValueError(f"{results_dir} is not a directory")
+
+        try:
+            self.summary_file_path = f"{results_dir}/{summary_filename}.json"
+            with open(self.summary_file_path, "w") as f:
+                json.dump(results.to_dict(), f, indent=4, default=str)
+        except Exception as e:
+            print(results.to_dict())
+            raise e
+
+        try:
+            self.individual_responses_file_path = f"{results_dir}/{individual_responses_filename}.json"
+            with open(self.individual_responses_file_path, "w") as f:
+                json.dump(individual_responses, f, indent=4)
+        except Exception as e:
+            print(individual_responses)
+            raise e
+
+
+class CustomPerformanceEvaluator(BasePerformanceEvaluator):
+    def __init__(
+        self, 
+        input_file_path: str,
+        *args, 
+        **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        self.file_name = os.path.basename(input_file_path)
+        self.dataset = self.read_dataset(input_file_path)
+        self.prompt_key = list(self.dataset[0].keys())[0]
+    
+    @staticmethod
+    def read_dataset(input_file_path: str):
+        """Utility function for reading in the `.jsonl` file provided by the user for custom dataset evaluation.
+        
+        Args:
+            input_file_path (str): The absolute file path of the input file provided by the user
+
+        Returns:
+            List[Dict]: A list of json objects (python dictionaries) containing the individual prompts the user wants to evaluate on
+        """
+        with open(input_file_path, 'r') as file:
+            data = [json.loads(line) for line in file]
+        return data
+
+    def create_output_filename(self):
+        """Utility for creating a unique filename for a custom benchmarking experiment with a dataset.
+
+        Returns:
+            str: Filename for the custom benchmark run.
+        """
+        return f"{self.model_name}_{self.file_name}_{self.num_workers}_{self.generation_mode}"
+
+    def run_benchmark(
+            self, 
+            sampling_params: Dict[str, Any]
```

Comment:
My logic here for keeping the sampling_params as a method argument instead of a class attribute is that the user will tend to alter params multiple times for a single dataset to see how it affects performance. For instance, they may try one run on their dataset with Temperature=0.1 and Top_K=15, and then they try again with new values, but the same dataset. Because of this, I separated the values that I believed would be changed more frequently so we could avoid instantiating multiple objects and taking up a bunch of local memory.

Reply:
fixed ✅